### DISCUSSION TITLE
[FEATURE] Minimum GasAdjustment API

### DISCUSSION THREAD

**Discussion by TheArchitect123** [classic-terra/core#183] [Ideas]
**Problem definition**
Can we have a minimum GasAdjustment (recommended value) that dApps can pull from the API, which will basically allow them to adjust the gas calculated to make sure that GasFees for the transaction always work. But it's a value that comes from the blockchain, and the chain is telling them how much the gas adjustment needs to be for a successful transaction (always). 

Because of the BurnTax for LUNC, users have to pay more, however sometimes the GasAdjustment (even if set to 2), it doesn't work. There's not enough Gas.
I have to manually set the Gas Adjustment to 3, to make sure the transaction always works, 2.5 is pushing it. 

![image](https://user-images.githubusercontent.com/20945125/222012475-82906c27-71e0-4d7a-a74a-01ef61634fa5.png)

**Comment by ZaradarBH** [comment]
To be honest I have been wondering a bit about some of these "out-of-gas" errors where the user almost has enough gas, but not quiet, and here is my take on it:

1) By and large validators use @ https://terra-classic-fcd.publicnode.com/v1/txs/gas_prices to figure out what their minimum gas price should be. However there is no enforcement of what the minimum value actually is from validator to validator, which means from time to time your TX might hit an "outlier".

2) Estimation of TX gas is managed via https://github.com/classic-terra/core/blob/main/custom/auth/client/utils/feeutils.go that is part of the terrad CLI. However its an oversimplistic method that happens on the "client-side" which means it is going to be prone to "offsets" due to factors such as time and "gamma" (rate-of-change-on-delta), which will cause the estimation to almost always be "wrong" once the actual TX is processed. Hence the need for the adjustment multiplier to ensure that the "offset" is always higher then the minimum requirement.

3) Considering 1 + 2, calculating the "desired GasAdjustment setting" on-chain is going to be a little tricky if we do not get validators to submit their individual min-gas-price settings and possibly the FCD gas prices as well. This would allow for on-chain calculation of a gas-adjustment-delta for the upcoming blocks which should work across all validators, granted that they do not decide to have fun and change these settings from block to block 🎪 🥳

**Reply by geoffmunn** [reply]
I've written a loop in the python project to check for code 11 responses, and bump up the gas adjustment value by 0.1. It keeps trying this until it either works or hits an upper limit set by the user.

It works well, but if you set the initial value far too low then it could take a very long time to eventually find a value that works.
